### PR TITLE
Issue #1878: Upgrade dependencies

### LINKED ISSUE #1878: Update dependencies to latest versions
**Problem description**
A lot of our dependencies are running very old versions. 

**Problem location**
gradle.properties

**Suggestions for an improvement**
We should update to the latest versions of these so that we can benefit from the latest updates to these packages.

### PR DESCRIPTION
**Pull request by tkaitchuck**
**Change log description**
Updates all of our dependencies to their latest published version.

**Purpose of the change**
Fixes #1878 

**What the code does**
Changes our the version numbers in gradle.properties, and makes the required code changes to build. (Auto-generated source has been re-generated.)

**How to verify it**
We will need to re-run all tests (including system and integration tests) to verify there have not been any subtle impacts.

### CODE REVIEW

File: controller/src/main/java/io/pravega/controller/server/rest/generated/api/ScopesApi.java [hunk 1]
```diff
@@ -27,18 +27,41 @@
 import org.glassfish.jersey.media.multipart.FormDataContentDisposition;
 import org.glassfish.jersey.media.multipart.FormDataParam;
 
+import javax.servlet.ServletConfig;
 import javax.ws.rs.core.Context;
 import javax.ws.rs.core.Response;
 import javax.ws.rs.core.SecurityContext;
 import javax.ws.rs.*;
+import javax.validation.constraints.*;
```
Comment (Prabha-Veerubhotla) [inline controller/src/main/java/io/pravega/controller/server/rest/generated/api/ScopesApi.java]: `*`  should be avoided in imports.

File: controller/src/main/java/io/pravega/controller/server/rest/generated/api/ScopesApiService.java [hunk 2]
```diff
@@ -24,20 +24,20 @@
 
 import javax.ws.rs.core.Response;
 import javax.ws.rs.core.SecurityContext;
-
+import javax.validation.constraints.*;
```
Comment (Prabha-Veerubhotla) [inline controller/src/main/java/io/pravega/controller/server/rest/generated/api/ScopesApiService.java]: same as above.

File: controller/src/main/java/io/pravega/controller/server/rest/generated/api/ScopesApiService.java [hunk 3]
```diff
@@ -24,20 +24,20 @@
...
-    public abstract Response getScalingEvents(String scopeName,String streamName,Long from,Long to,SecurityContext securityContext) throws NotFoundException;
+    public abstract Response getScalingEvents(String scopeName,String streamName, @NotNull Long from, @NotNull Long to,SecurityContext securityContext) throws NotFoundException;
     public abstract Response getScope(String scopeName,SecurityContext securityContext) throws NotFoundException;
     public abstract Response getStream(String scopeName,String streamName,SecurityContext securityContext) throws NotFoundException;
     public abstract Response listReaderGroups(String scopeName,SecurityContext securityContext) throws NotFoundException;
     public abstract Response listScopes(SecurityContext securityContext) throws NotFoundException;
-    public abstract Response listStreams(String scopeName,String showInternalStreams,SecurityContext securityContext) throws NotFoundException;
+    public abstract Response listStreams(String scopeName, String showInternalStreams,SecurityContext securityContext) throws NotFoundException;
```
Comment (Prabha-Veerubhotla) [inline controller/src/main/java/io/pravega/controller/server/rest/generated/api/ScopesApiService.java]: please fix this unintended space.

File: controller/src/main/java/io/pravega/controller/server/rest/generated/api/impl/ScopesApiServiceImpl.java [hunk 4]
```diff
@@ -24,7 +24,7 @@
 
 import javax.ws.rs.core.Response;
 import javax.ws.rs.core.SecurityContext;
-
+import javax.validation.constraints.*;
```
Comment (Prabha-Veerubhotla) [inline controller/src/main/java/io/pravega/controller/server/rest/generated/api/impl/ScopesApiServiceImpl.java]: same as above.

File: controller/src/main/java/io/pravega/controller/server/rest/generated/api/impl/ScopesApiServiceImpl.java [hunk 5]
```diff
@@ -78,7 +78,7 @@ public Response listScopes(SecurityContext securityContext) throws NotFoundExcep
         return Response.ok().entity(new ApiResponseMessage(ApiResponseMessage.OK, "magic!")).build();
     }
     @Override
-    public Response listStreams(String scopeName, String showInternalStreams, SecurityContext securityContext) throws NotFoundException {
+    public Response listStreams(String scopeName,  String showInternalStreams, SecurityContext securityContext) throws NotFoundException {
```
Comment (Prabha-Veerubhotla) [inline controller/src/main/java/io/pravega/controller/server/rest/generated/api/impl/ScopesApiServiceImpl.java]: space.

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/CreateStreamRequest.java [hunk 6]
```diff
@@ -7,30 +20,32 @@
 import io.pravega.controller.server.rest.generated.model.ScalingConfig;
 import io.swagger.annotations.ApiModel;
 import io.swagger.annotations.ApiModelProperty;
-
-
-
+import javax.validation.constraints.*;
```
Comment (Prabha-Veerubhotla) [inline controller/src/main/java/io/pravega/controller/server/rest/generated/model/CreateStreamRequest.java]: `*` imports.

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/ReaderGroupProperty.java [hunk 7]
```diff
@@ -7,32 +20,35 @@
 import io.swagger.annotations.ApiModelProperty;
 import java.util.ArrayList;
 import java.util.List;
-
-
-
+import javax.validation.constraints.*;
```
Comment (Prabha-Veerubhotla) [inline controller/src/main/java/io/pravega/controller/server/rest/generated/model/ReaderGroupProperty.java]: `*` imports.

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/ReaderGroupsList.java [hunk 8]
```diff
@@ -8,31 +21,34 @@
 import io.swagger.annotations.ApiModelProperty;
 import java.util.ArrayList;
 import java.util.List;
-
-
-
+import javax.validation.constraints.*;
```
Comment (Prabha-Veerubhotla) [inline controller/src/main/java/io/pravega/controller/server/rest/generated/model/ReaderGroupsList.java]: same as above.

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/ReaderGroupsListReaderGroups.java [hunk 9]
```diff
@@ -1,30 +1,43 @@
...
 
 import java.util.Objects;
 import com.fasterxml.jackson.annotation.JsonProperty;
 import com.fasterxml.jackson.annotation.JsonCreator;
 import io.swagger.annotations.ApiModel;
 import io.swagger.annotations.ApiModelProperty;
-
-
-
+import javax.validation.constraints.*;
```
Comment (Prabha-Veerubhotla) [inline controller/src/main/java/io/pravega/controller/server/rest/generated/model/ReaderGroupsListReaderGroups.java]: same as above.

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/RetentionConfig.java [hunk 10]
```diff
@@ -6,9 +19,7 @@
 import com.fasterxml.jackson.annotation.JsonValue;
 import io.swagger.annotations.ApiModel;
 import io.swagger.annotations.ApiModelProperty;
-
-
-
+import javax.validation.constraints.*;
```
Comment (Prabha-Veerubhotla) [inline controller/src/main/java/io/pravega/controller/server/rest/generated/model/RetentionConfig.java]: same as above.

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/ScaleMetadata.java [hunk 11]
```diff
@@ -8,28 +21,29 @@
 import io.swagger.annotations.ApiModelProperty;
 import java.util.ArrayList;
 import java.util.List;
-
-
-
+import javax.validation.constraints.*;
```
Comment (Prabha-Veerubhotla) [inline controller/src/main/java/io/pravega/controller/server/rest/generated/model/ScaleMetadata.java]: same as above.

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/ScalingConfig.java [hunk 12]
```diff
@@ -6,9 +19,7 @@
 import com.fasterxml.jackson.annotation.JsonValue;
 import io.swagger.annotations.ApiModel;
 import io.swagger.annotations.ApiModelProperty;
-
-
-
+import javax.validation.constraints.*;
```
Comment (Prabha-Veerubhotla) [inline controller/src/main/java/io/pravega/controller/server/rest/generated/model/ScalingConfig.java]: same as above.

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/ScalingEventList.java [hunk 13]
```diff
@@ -8,31 +21,34 @@
 import io.swagger.annotations.ApiModelProperty;
 import java.util.ArrayList;
 import java.util.List;
-
-
-
+import javax.validation.constraints.*;
```
Comment (Prabha-Veerubhotla) [inline controller/src/main/java/io/pravega/controller/server/rest/generated/model/ScalingEventList.java]: same as above.

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/ScopeProperty.java [hunk 14]
```diff
@@ -1,30 +1,43 @@
...
 
 import java.util.Objects;
 import com.fasterxml.jackson.annotation.JsonProperty;
 import com.fasterxml.jackson.annotation.JsonCreator;
 import io.swagger.annotations.ApiModel;
 import io.swagger.annotations.ApiModelProperty;
-
-
-
+import javax.validation.constraints.*;
```
Comment (Prabha-Veerubhotla) [inline controller/src/main/java/io/pravega/controller/server/rest/generated/model/ScopeProperty.java]: same as above.

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/ScopesList.java [hunk 15]
```diff
@@ -8,31 +21,34 @@
 import io.swagger.annotations.ApiModelProperty;
 import java.util.ArrayList;
 import java.util.List;
-
-
-
+import javax.validation.constraints.*;
```
Comment (Prabha-Veerubhotla) [inline controller/src/main/java/io/pravega/controller/server/rest/generated/model/ScopesList.java]: same as above.

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/Segment.java [hunk 16]
```diff
@@ -1,36 +1,52 @@
...
 
 import java.util.Objects;
 import com.fasterxml.jackson.annotation.JsonProperty;
 import com.fasterxml.jackson.annotation.JsonCreator;
 import io.swagger.annotations.ApiModel;
 import io.swagger.annotations.ApiModelProperty;
-
-
-
+import javax.validation.constraints.*;
```
Comment (Prabha-Veerubhotla) [inline controller/src/main/java/io/pravega/controller/server/rest/generated/model/Segment.java]: same as above.

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/StreamProperty.java [hunk 17]
```diff
@@ -7,32 +20,35 @@
 import io.pravega.controller.server.rest.generated.model.ScalingConfig;
 import io.swagger.annotations.ApiModel;
 import io.swagger.annotations.ApiModelProperty;
-
-
-
+import javax.validation.constraints.*;
```
Comment (Prabha-Veerubhotla) [inline controller/src/main/java/io/pravega/controller/server/rest/generated/model/StreamProperty.java]: same as above.

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/StreamState.java [hunk 18]
```diff
@@ -6,9 +19,7 @@
 import com.fasterxml.jackson.annotation.JsonValue;
 import io.swagger.annotations.ApiModel;
 import io.swagger.annotations.ApiModelProperty;
-
-
-
+import javax.validation.constraints.*;
```
Comment (Prabha-Veerubhotla) [inline controller/src/main/java/io/pravega/controller/server/rest/generated/model/StreamState.java]: same as above.

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/UpdateStreamRequest.java [hunk 19]
```diff
@@ -7,28 +20,29 @@
 import io.pravega.controller.server.rest.generated.model.ScalingConfig;
 import io.swagger.annotations.ApiModel;
 import io.swagger.annotations.ApiModelProperty;
-
-
-
+import javax.validation.constraints.*;
```
Comment (Prabha-Veerubhotla) [inline controller/src/main/java/io/pravega/controller/server/rest/generated/model/UpdateStreamRequest.java]: same as above.

File: gradle.properties [hunk 20]
```diff
@@ -11,37 +11,36 @@
 dockerExecutable=/usr/bin/docker
 
 #3rd party Versions
-apacheCuratorVersion=2.11.0
-checkstyleToolVersion=7.1
+apacheCuratorVersion=4.0.+
+checkstyleToolVersion=8.2
```
Comment (Prabha-Veerubhotla) [inline gradle.properties]: Can you also make these alphabetic ?

File: controller/src/main/java/io/pravega/controller/server/rest/generated/api/RFC3339DateFormat.java [hunk 21]
```diff
@@ -0,0 +1,19 @@
...
+
+    // Same as ISO8601DateFormat but serializing milliseconds.
+    @Override
+    public StringBuffer format(Date date, StringBuffer toAppendTo, FieldPosition fieldPosition) {
+        String value = ISO8601Utils.format(date, true);
+        toAppendTo.append(value);
+        return toAppendTo;
+    }
+
+}
```
[the remark this answers is not in the document]
Comment (tkaitchuck) [reply controller/src/main/java/io/pravega/controller/server/rest/generated/api/RFC3339DateFormat.java]: This and all the below are generated code. As they are not human created, we should avoid editing them so that we can always re-create them when a new version is released.

File: controller/src/main/java/io/pravega/controller/server/rest/generated/api/ScopesApiService.java (hunk 2 above)
Comment (tkaitchuck) [reply controller/src/main/java/io/pravega/controller/server/rest/generated/api/ScopesApiService.java]: Generated code

File: controller/src/main/java/io/pravega/controller/server/rest/generated/api/ScopesApi.java (hunk 1 above)
Comment (tkaitchuck) [reply controller/src/main/java/io/pravega/controller/server/rest/generated/api/ScopesApi.java]: Generated code

File: controller/src/main/java/io/pravega/controller/server/rest/generated/api/ScopesApiService.java (hunk 3 above)
Comment (tkaitchuck) [reply controller/src/main/java/io/pravega/controller/server/rest/generated/api/ScopesApiService.java]: Generated code

File: controller/src/main/java/io/pravega/controller/server/rest/generated/api/impl/ScopesApiServiceImpl.java (hunk 4 above)
Comment (tkaitchuck) [reply controller/src/main/java/io/pravega/controller/server/rest/generated/api/impl/ScopesApiServiceImpl.java]: Generated code

File: controller/src/main/java/io/pravega/controller/server/rest/generated/api/impl/ScopesApiServiceImpl.java (hunk 5 above)
Comment (tkaitchuck) [reply controller/src/main/java/io/pravega/controller/server/rest/generated/api/impl/ScopesApiServiceImpl.java]: Generated code

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/CreateStreamRequest.java (hunk 6 above)
Comment (tkaitchuck) [reply controller/src/main/java/io/pravega/controller/server/rest/generated/model/CreateStreamRequest.java]: Generated code

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/ReaderGroupProperty.java (hunk 7 above)
Comment (tkaitchuck) [reply controller/src/main/java/io/pravega/controller/server/rest/generated/model/ReaderGroupProperty.java]: Generated code

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/ReaderGroupsList.java (hunk 8 above)
Comment (tkaitchuck) [reply controller/src/main/java/io/pravega/controller/server/rest/generated/model/ReaderGroupsList.java]: Generated code

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/ReaderGroupsListReaderGroups.java (hunk 9 above)
Comment (tkaitchuck) [reply controller/src/main/java/io/pravega/controller/server/rest/generated/model/ReaderGroupsListReaderGroups.java]: Generated code

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/RetentionConfig.java (hunk 10 above)
Comment (tkaitchuck) [reply controller/src/main/java/io/pravega/controller/server/rest/generated/model/RetentionConfig.java]: Generated code

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/ScaleMetadata.java (hunk 11 above)
Comment (tkaitchuck) [reply controller/src/main/java/io/pravega/controller/server/rest/generated/model/ScaleMetadata.java]: Generated code

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/ScalingConfig.java (hunk 12 above)
Comment (tkaitchuck) [reply controller/src/main/java/io/pravega/controller/server/rest/generated/model/ScalingConfig.java]: Generated code

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/ScalingEventList.java (hunk 13 above)
Comment (tkaitchuck) [reply controller/src/main/java/io/pravega/controller/server/rest/generated/model/ScalingEventList.java]: Generated code

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/ScopesList.java (hunk 15 above)
Comment (tkaitchuck) [reply controller/src/main/java/io/pravega/controller/server/rest/generated/model/ScopesList.java]: Generated code

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/ScopeProperty.java (hunk 14 above)
Comment (tkaitchuck) [reply controller/src/main/java/io/pravega/controller/server/rest/generated/model/ScopeProperty.java]: Generated code

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/Segment.java (hunk 16 above)
Comment (tkaitchuck) [reply controller/src/main/java/io/pravega/controller/server/rest/generated/model/Segment.java]: Generated code

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/StreamProperty.java (hunk 17 above)
Comment (tkaitchuck) [reply controller/src/main/java/io/pravega/controller/server/rest/generated/model/StreamProperty.java]: Generated code

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/StreamState.java (hunk 18 above)
Comment (tkaitchuck) [reply controller/src/main/java/io/pravega/controller/server/rest/generated/model/StreamState.java]: Generated code

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/UpdateStreamRequest.java (hunk 19 above)
Comment (tkaitchuck) [reply controller/src/main/java/io/pravega/controller/server/rest/generated/model/UpdateStreamRequest.java]: Generated code

File: gradle.properties (hunk 20 above)
Comment (tkaitchuck) [reply gradle.properties]: Done

File: client/src/main/java/io/pravega/client/netty/impl/ConnectionFactoryImpl.java [hunk 22]
```diff
@@ -66,7 +66,7 @@ public ConnectionFactoryImpl(boolean ssl) {
         this.ssl = ssl;
         try {
             this.group = new EpollEventLoopGroup();
-        } catch (ExceptionInInitializerError | NoClassDefFoundError e) {
+        } catch (ExceptionInInitializerError | UnsatisfiedLinkError | NoClassDefFoundError e) {
```
Comment (fpj) [inline client/src/main/java/io/pravega/client/netty/impl/ConnectionFactoryImpl.java]: This change doesn't belong here, let's do it elsewhere, please.

File: gradle.properties [hunk 23]
```diff
@@ -11,37 +11,36 @@
 dockerExecutable=/usr/bin/docker
 
 #3rd party Versions
-apacheCuratorVersion=2.11.0
-checkstyleToolVersion=7.1
-findbugsVersion=3.0.1
+apacheCommonsCsvVersion=1.5
+apacheCuratorVersion=4.0.+
```
Comment (fpj) [inline gradle.properties]: Apparently the use of `+` for version format causes issues when publishing to maven central, see it in PR #1929. Let's use the full version instead.

Comment (fpj) [reply gradle.properties]: There are a bunch of uses of `+` here, please fix all those.

File: test/system/src/main/java/io/pravega/test/system/framework/services/PravegaControllerService.java [hunk 24]
```diff
@@ -140,7 +140,7 @@ private App createPravegaControllerApp() {
                 setSystemProperty("REST_SERVER_PORT", String.valueOf(REST_PORT)) +
                 setSystemProperty("log.level", "DEBUG") +
                 setSystemProperty("curator-default-session-timeout", String.valueOf(10 * 1000));
-        Map<String, String> map = new HashMap<>();
+        Map<String, Object> map = new HashMap<>();
```
Comment (fpj) [inline test/system/src/main/java/io/pravega/test/system/framework/services/PravegaControllerService.java]: There a few changes from `String` to `Object`, is this related to any of the version upgrades?

File: shared/src/test/java/io/pravega/shared/segment/SegmentToContainerMapperTests.java [hunk 25]
```diff
@@ -42,13 +42,13 @@ public void testConstructor() {
...
     @Test
     public void testUniformMapping() {
-        testUniformMapping(20, 10000, 0.10, (num) -> getSegmentName(num));
-        testUniformMapping(20, 10000, 0.10, (num) -> Integer.toString(num));
-        testUniformMapping(100, 100000, 0.10, (num) -> getSegmentName(num));
-        testUniformMapping(100, 100000, 0.10, (num) -> Integer.toString(num));
-        testUniformMapping(100, 100000, 0.10, (num) -> Integer.toBinaryString(num));
-        testUniformMapping(100, 100000, 0.10, (num) -> Integer.toOctalString(num));
-        testUniformMapping(100, 100000, 0.10, (num) -> Integer.toHexString(num));
+        testUniformMapping(20, 10000, 0.10, num -> getSegmentName(num));
```
Comment (fpj) [inline shared/src/test/java/io/pravega/shared/segment/SegmentToContainerMapperTests.java]: This change does not seem to be related to the upgrades proposed here, if that's the case and for the future, let's try to avoid unrelated formatting changes.

**Comment by fpj** [comment]
Thanks for doing this version upgrade, @tkaitchuck . I have left a few comments in the PR. Also, there is one conflict and the last build had a test failure that seems to be related to the changes here:

```
io.pravega.segmentstore.server.host.HDFSIntegrationTest > testEndToEnd FAILED
    java.lang.NoSuchMethodError: com.google.common.base.Objects.toStringHelper(Ljava/lang/Object;)Lcom/google/common/base/Objects$ToStringHelper;
        at org.apache.hadoop.metrics2.lib.MetricsRegistry.toString(MetricsRegistry.java:419)
        at java.lang.String.valueOf(String.java:2994)
        at java.lang.StringBuilder.append(StringBuilder.java:131)
        at org.apache.hadoop.ipc.metrics.RpcMetrics.<init>(RpcMetrics.java:74)
        at org.apache.hadoop.ipc.metrics.RpcMetrics.create(RpcMetrics.java:80)
        at org.apache.hadoop.ipc.Server.<init>(Server.java:2590)
        at org.apache.hadoop.ipc.RPC$Server.<init>(RPC.java:958)
        at org.apache.hadoop.ipc.ProtobufRpcEngine$Server.<init>(ProtobufRpcEngine.java:366)
        at org.apache.hadoop.ipc.ProtobufRpcEngine.getServer(ProtobufRpcEngine.java:341)
        at org.apache.hadoop.ipc.RPC$Builder.build(RPC.java:800)
        at org.apache.hadoop.hdfs.server.namenode.NameNodeRpcServer.<init>(NameNodeRpcServer.java:421)
        at org.apache.hadoop.hdfs.server.namenode.NameNode.createRpcServer(NameNode.java:755)
        at org.apache.hadoop.hdfs.server.namenode.NameNode.initialize(NameNode.java:697)
        at org.apache.hadoop.hdfs.server.namenode.NameNode.<init>(NameNode.java:898)
        at org.apache.hadoop.hdfs.server.namenode.NameNode.<init>(NameNode.java:877)
        at org.apache.hadoop.hdfs.server.namenode.NameNode.createNameNode(NameNode.java:1603)
        at org.apache.hadoop.hdfs.MiniDFSCluster.createNameNode(MiniDFSCluster.java:1158)
        at org.apache.hadoop.hdfs.MiniDFSCluster.createNameNodesAndSetConf(MiniDFSCluster.java:1033)
        at org.apache.hadoop.hdfs.MiniDFSCluster.initMiniDFSCluster(MiniDFSCluster.java:826)
        at org.apache.hadoop.hdfs.MiniDFSCluster.<init>(MiniDFSCluster.java:483)
        at org.apache.hadoop.hdfs.MiniDFSCluster$Builder.build(MiniDFSCluster.java:442)
        at io.pravega.segmentstore.storage.impl.hdfs.HDFSClusterHelpers.createMiniDFSCluster(HDFSClusterHelpers.java:32)
        at io.pravega.segmentstore.server.host.HDFSIntegrationTest.setUp(HDFSIntegrationTest.java:51)
```

File: client/src/main/java/io/pravega/client/netty/impl/ConnectionFactoryImpl.java (hunk 22 above)
Comment (tkaitchuck) [reply client/src/main/java/io/pravega/client/netty/impl/ConnectionFactoryImpl.java]: Yes, netty changed the exception it throws.

File: shared/src/test/java/io/pravega/shared/segment/SegmentToContainerMapperTests.java (hunk 25 above)
Comment (tkaitchuck) [reply shared/src/test/java/io/pravega/shared/segment/SegmentToContainerMapperTests.java]: It is related, as are the similar changes above. The new version of checkstyle considers it an error to surround  a single parameter in a lambda with parenthesis. (I don't know why, but it does.) So I'm removing the parenthesis

File: test/system/src/main/java/io/pravega/test/system/framework/services/PravegaControllerService.java (hunk 24 above)
Comment (tkaitchuck) [reply test/system/src/main/java/io/pravega/test/system/framework/services/PravegaControllerService.java]: Yes. The type on the curator interface changed.

File: controller/src/main/java/io/pravega/controller/store/checkpoint/ZKCheckpointStore.java [hunk 26]
```diff
@@ -312,7 +312,8 @@ private void updateNode(String path, byte[] data) throws CheckpointStoreExceptio
         } catch (KeeperException.NoNodeException e) {
             // Return empty list for KeeperException.NoNodeException.
             return Collections.emptyList();
-        } catch (KeeperException.ConnectionLossException | KeeperException.OperationTimeoutException e) {
+        } catch (KeeperException.ConnectionLossException | KeeperException.OperationTimeoutException
+                | KeeperException.SessionExpiredException e) {
```
Comment (fpj) [inline controller/src/main/java/io/pravega/controller/store/checkpoint/ZKCheckpointStore.java]: It sounds right that we should give a proper treatment to session expiration, but I have not looked close enough at this flow to know whether simply adding an OR is correct. Also, it sounds like other operations above are not handling it explicitly either. I suggest we don't do this change and create an issue for it. Ask @shiveshr to have a look.

Comment (tkaitchuck) [reply controller/src/main/java/io/pravega/controller/store/checkpoint/ZKCheckpointStore.java]: Without this change the tests will fail on the new BK. I don't know why this exception was changed. I'll talk with @shiveshr

Comment (shiveshr) [reply controller/src/main/java/io/pravega/controller/store/checkpoint/ZKCheckpointStore.java]: this looks right. Though should be redundant. 

Note: curator client automatically tries to reconnect internally for connection failures. However if there is a session expiry, we shutdown our curator client, which in turn results on controller instance being shutdown too. 
This is to make sure that the controller does not erroneously assume it has "lock" or is the "leader" and continues to perform other metadata updates which could happen if we were to silently create a new session if the previous session expired. 

So even though in each of the stores (like metadata store, host index store etc) we have translated sessionExpiry to connection loss exception, the underlying curator client is never asked to create a new session. 
We have a watcher on curator's connection state. And if it moves to sessionExpired, we stop controller service. 

So translating keeper.sessionExpired exception to connection loss exception makes controller's workflows treat this as a temporary connection loss as it retries/responds to the caller with connection loss failure. [Note: once session is expired, controller cannot write to zk for any metadata update. So this translation of exception is primarily targeted towards responding to caller with appropriate "connection loss" exception] 

Also, the controller instance will shutdown in the background because of session expiry.

**Comment by tkaitchuck** [comment]
@fpj  Fixed.
